File: trunk/java/SR-RBD/src/ddb/db/DbConnectorImpl.java
/*
 * To change this template, choose Tools | Templates
 * and open the template in the editor.
 */

package ddb.db;

import java.io.*;
import java.sql.*;

import ddb.Config;
import ddb.Logger;

/**
 *
 * @author xeonic
 */
public class DbConnectorImpl implements DbConnector
{
    private Connection connection;
    private static final String LOGGING_NAME = "DbConnector";
    

    private static DbConnector instance = null;
    
    String URL_BEGIN = "jdbc:mysql";
    String DRIVER = "com.mysql.jdbc.Driver";

    private DbConnectorImpl()
    {
        String url = URL_BEGIN + "://" + Config.Host() + "/" + Config.Database();

        try
        {
            Class.forName(DRIVER).newInstance();
            connection = DriverManager.getConnection(url, Config.User(), Config.Password());
        }
        catch (InstantiationException ex)
        {
            Logger.getInstance().log("Can't connect to the database: " + ex.getMessage(),
                    LOGGING_NAME, Logger.Level.SEVERE);
        }
        catch (IllegalAccessException ex)
        {
            Logger.getInstance().log("Can't connect to the database: " + ex.getMessage(),
                    LOGGING_NAME, Logger.Level.SEVERE);
        }
        catch (ClassNotFoundException ex)
        {
            Logger.getInstance().log("Can't connect to the database: " + ex.getMessage(),
                    LOGGING_NAME, Logger.Level.SEVERE);
        }
        catch (SQLException ex)
        {
            Logger.getInstance().log("Can't connect to the database: " + ex.getMessage(),
                    LOGGING_NAME, Logger.Level.SEVERE);

        }
    }

    public static DbConnector getInstance()
    {
    	if( instance == null )
    		instance = new DbConnectorImpl();
    	
        return instance;
    }

    public DatabaseTable query(String queryString) throws DBException
    {
    	Statement statement = null;
    	DatabaseTable result = null;
        try
        {
        	statement = connection.createStatement();
			statement.execute(queryString);
			result = new DatabaseTable(statement.getResultSet());
		} 
        catch (SQLException e) 
        {
        	throw new DBException(e);
		}
        return result;
    }
    
    public void clearDatabase() throws DBException
    {                                             
        Statement statement = null;               
        ResultSet result = null;                  
        try                                       
        {                                         
            statement = connection.createStatement();
            statement.execute("SHOW TABLES");        

            Logger.getInstance().log("Clearing database...", LOGGING_NAME,
                    Logger.Level.INFO);                                   

            result = statement.getResultSet();

            while(result.next())
            {                   
               statement = connection.createStatement();
                statement.execute("DROP TABLE " + result.getString(1));
                Logger.getInstance().log("DROP TABLE: " + result.getString(1),
                        LOGGING_NAME, Logger.Level.INFO);                     
            }                                                                 
        }                                                                     
        catch (SQLException e)                                                
        {                                                                     
            throw new DBException(e);                                         
        }                                                                     
    }                                                                         

    public String dumpTable(String tableName) throws DumpTableException
    {                                        
        String output = "";                  
        String cmd = "mysqldump -u " + Config.User() + " -p" + Config.Password() + " " + 
        				Config.Database() + " " + tableName;                                             

        Logger.getInstance().log(cmd, LOGGING_NAME, Logger.Level.INFO);

        Process p = null;
        try              
        {                
            p = Runtime.getRuntime().exec(cmd);
        }                                      
        catch (IOException e)                  
        {                                      
            Logger.getInstance().log("IO error during table dump: " + e.getMessage(),
                    LOGGING_NAME, Logger.Level.WARNING); 
            throw new DumpTableException(e);
        }                                                                            

        BufferedReader input = new BufferedReader(new InputStreamReader(p.getInputStream()));

        try
        {  
            String line = null;
            while ((line = input.readLine()) != null)
            {                                        
                if(!line.startsWith("/*") && !line.startsWith("--"))
                    output += line;                                 
            }                                                       
        }                                                           
        catch (IOException e)                                       
        {                                                           
            Logger.getInstance().log("IO error during table dump: " + e.getMessage(),
                    LOGGING_NAME, Logger.Level.WARNING);
            throw new DumpTableException(e);
        }                                                                            

        return output;
    }                 

    public void importTable(String tableName, int version, String dump) throws ImportTableException
    {                                   
        String cmd = "mysql -u " + Config.User() + " -p" + Config.Password() + " " +
        				Config.Database();

        Logger.getInstance().log(cmd, LOGGING_NAME, Logger.Level.INFO);

        Process p = null;
        try
        {
            p = Runtime.getRuntime().exec(cmd);
            OutputStream stdin = p.getOutputStream();
            stdin.write(dump.toString().getBytes());
            stdin.flush();
        }
        catch (IOException e)
        {
            Logger.getInstance().log("IO error during table import: " + e.getMessage(),
                    LOGGING_NAME, Logger.Level.WARNING);
            throw new ImportTableException(e);
        }
        DatabaseStateImpl.getInstance().setTableVersion(tableName, version);
    }

}

File: trunk/C#/RBD/RBD/DB/DbConnectorImpl.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Diagnostics;
using System.IO;
using MySql.Data.MySqlClient;

namespace RBD.DB
{
    public class DbConnectorImpl : DbConnector
    {
        static DbConnector instance = new DbConnectorImpl();
        const String LOGGING_NAME = "SqlConnector";

        String server;
        String database;
        String username;
        String password;

        MySqlConnection connection;

        DbConnectorImpl()
        {
            server = Config.Host();
            database = Config.Database();
            username = Config.User();
            password = Config.Password();
        }

        void openDatabase()
        {
            try
            {
                string MyConString = "SERVER=" + server +
                                     ";DATABASE=" + database +
                                     ";UID=" + username +
                                     ";PASSWORD=" + password + ";";
                Logger.getInstance().log(MyConString, LOGGING_NAME, Logger.Level.INFO);
                connection = new MySqlConnection(MyConString);
                connection.Open();
            }
            catch (MySqlException e)
            {
                Logger.getInstance().log("Could not open database: " + e.Message,
                    LOGGING_NAME, Logger.Level.SEVERE);
            }
        }

        public static DbConnector getInstance()
        {
            return instance;
        }

        public DatabaseTable query(String queryString)
        {
            if (connection == null)
                openDatabase();

            MySqlDataReader Reader = null;
            if (connection.State == System.Data.ConnectionState.Open)
            {

                MySqlCommand command = connection.CreateCommand();
                command.CommandText = queryString;
                Reader = command.ExecuteReader();
            }
            return new DatabaseTable(Reader);
        }

        public void clearDatabase()
        {
            if (connection == null)
                openDatabase();

            MySqlDataReader reader = null;
            Logger.getInstance().log("Clearing database...", "SqlConnector", Logger.Level.INFO);
            if (connection.State == System.Data.ConnectionState.Open)
            {
                MySqlCommand command = connection.CreateCommand();
                command.CommandText = "SHOW TABLES";
                reader = command.ExecuteReader();
            }
            List<String> tables = new List<string>();
            while(reader.Read())
                tables.Add(reader.GetValue(0).ToString());
            
            reader.Close();

            if (connection.State == System.Data.ConnectionState.Open)
            {
                foreach (String tableName in tables)
                {
                    MySqlCommand command = connection.CreateCommand();
                    command.CommandText = "DROP TABLE " + tableName;
                    Logger.getInstance().log(command.CommandText, "SqlConnector", Logger.Level.INFO);
                    MySqlDataReader reader2 = command.ExecuteReader();
                    reader2.Close();
                }
            }
        }

        public String dumpTable(String tableName)
        {
            String output = "";
            ProcessStartInfo mysqldump = new ProcessStartInfo();
            mysqldump.FileName = "mysqldump";
            //mysqldump.Arguments = "-u " + USERNAME + " -p" + PASSWORD + " " + DATABASE + " " + tableName;
            mysqldump.Arguments = "-u " + username + " -p" + password + " -h " + server + " " + database + " " + tableName;
            mysqldump.UseShellExecute = false;
            mysqldump.RedirectStandardOutput = true;

            Logger.getInstance().log(mysqldump.FileName + " " + mysqldump.Arguments, LOGGING_NAME, Logger.Level.INFO);
            
            using (Process process = Process.Start(mysqldump))
            {
                using (StreamReader reader = process.StandardOutput)
                {
                    while (!reader.EndOfStream)
                    {
                        string line = reader.ReadLine();
                        if (!line.StartsWith("/*") && !line.StartsWith("--"))
                            output += line.Trim();
                    }
                }
            }
            
            return output;
        }

        public void importTable(string tableName, int version, String dump)
        {
            ProcessStartInfo mysql = new ProcessStartInfo();
            mysql.FileName = "mysql";
            //mysql.Arguments = "-u " + USERNAME + " -p" + PASSWORD + " " + DATABASE;
            mysql.Arguments = "-u " + username + " -p" + password + " -h " + server + " " + database;
            mysql.UseShellExecute = false;
            mysql.RedirectStandardInput = true;

            Logger.getInstance().log(mysql.FileName + " " + mysql.Arguments, LOGGING_NAME, Logger.Level.INFO);

            using (Process process = Process.Start(mysql))
                using (StreamWriter writer = process.StandardInput)
                   writer.Write(dump);
            DatabaseStateImpl.getInstance().setTableVersion(tableName, version);

        }
    }
}
